fix(curriculum): removed global variable ab(use) in lunch picker (#58014)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-lunch-picker-program/66db529d37ad966480ebb633.md

1. You should create a `lunches` variable and assign it an empty array that will be used to store lunch items.

2. You should create a function `addLunchToEnd` that takes an array as the first argument and a string as the second argument. The function should:
    - Add the string to the end of the array.

>     - Log the string `"[Lunch Item] added to the end of the lunch menu."` to the console.

Reviewer: ```suggestion
    - Log the string `"[Lunch Item] added to the end of the lunch menu."` to the console where `[Lunch Item]` is the string passed to the function.
```

maybe it should specidied also in the other places, but here it's a must otherwise it's confusing

    - Return the updated array.

3. You should create a function `addLunchToStart` that takes an array as the first argument and a string as the second argument. The function should:
    - Add the string to the start of the array.

>     - Log the string `"[Lunch Item] added to the start of the lunch menu."` to the console.

Reviewer: ```suggestion
    - Log the string `"[Lunch Item] added to the start of the lunch menu."` to the console where `[Lunch Item]` is the string passed to the function.
```

    - Return the updated array.

4. You should create a function `removeLastLunch` that takes an array as its argument. The function should:
   - Remove the last lunch item from the array.

>    - If  the removal is successful, log the string `"[Lunch Item] removed from the end of the lunch menu."` to the console.

Reviewer: ```suggestion
   - If  the removal is successful, log the string `"[Lunch Item] removed from the end of the lunch menu."` to the console where `[Lunch Item]` is the element removed from the array.
```

   - If the array is empty, log the string `"No lunches to remove."` to the console.
   - Return the updated array.

5. You should create a function `removeFirstLunch` that takes an array as its argument. The function should:
   - Remove the first lunch item from the array.

>    - If the removal is successful, log the string `"[Lunch Item] removed from the start of the lunch menu."` to the console.

Reviewer: ```suggestion
   - If the removal is successful, log the string `"[Lunch Item] removed from the start of the lunch menu."` to the console where `[Lunch Item]` is the element removed from the array.
```

   - If the array is empty, log the string `"No lunches to remove."` to the console.
   - Return the updated array.

6. You should create a function `getRandomLunch` that takes an array as its argument. The function should:

>    - Select a random lunch item from the array.

Reviewer: ```suggestion
   - Select a random item from the array.
```

Reviewer: the items or elements in an array can be called both ways, considering the position in the curriculum there should not be any confusion, but if you prefer one over the other feel free to use only one

>    - If successful, log the string `"Randomly selected lunch: [Lunch Item]"` to the console.

Reviewer: ```suggestion
   - If successful, log the string `"Randomly selected lunch: [Lunch Item]"` to the console where `[Lunch Item]` is the random item from the array.
```

   - If the array is empty, log the string `"No lunches available."` to the console.

7. You should create a function `showLunchMenu` that takes an array as its argument and:

>    - If there are lunch items in the array, logs the string `"Menu items: [Lunch Item], [Lunch Item]...` to the console.

Reviewer: ```suggestion
   - If there are items in the array, logs the string `"Menu items: [Lunch Item], [Lunch Item]...` to the console where each `[Lunch item]` is one of the elements in the array, in order.
```

assert.isFunction(addLunchToEnd);
```

The function `addLunchToEnd` should take 2 arguments.

```js
assert.lengthOf(addLunchToEnd, 2);
```

`addLunchToEnd(lunches, "Tacos")` should log the string `"Tacos added to the end of the lunch menu."`.

```js

> const tempArr = [];
> const temp = console.log;
> try {
>   console.log = obj => tempArr.push(obj);
>   addLunchToEnd(lunches, "Pizza");
>   assert.strictEqual(tempArr[0], "Pizza added to the end of the lunch menu.");
> } finally {
>   console.log = temp;
> }

Reviewer 2: ```suggestion
const tempArr = [];
const temp = console.log;
const testLunches = []
try {
  console.log = obj => tempArr.push(obj);
  addLunchToEnd(testLunches, "Tacos");
  assert.strictEqual(tempArr[0], "Tacos added to the end of the lunch menu.");
  addLunchToEnd(testLunches, "Pizza");
  assert.strictEqual(tempArr[1], "Pizza added to the end of the lunch menu.");
} finally {
  console.log = temp;
}
```
We should test what's in the hint. Also, don't pass `lunches` to test the functions. It's better to create another array named differently to avoid confusion.

```

> The function `addLunchToEnd` should return the correct array.

Reviewer 2: ```suggestion
`addLunchToEnd(["Pizza", "Tacos"], "Burger")` should return `["Pizza", "Tacos", "Burger"]`.
```
I would change the other hints to something like this so one can see the expected result.

assert.isFunction(addLunchToEnd);
```

The function `addLunchToEnd` should have two parameters.

```js
assert.lengthOf(addLunchToEnd, 2);
```

`addLunchToEnd(lunches, "Tacos")` should log the string `"Tacos added to the end of the lunch menu."` to the console.

```js
const tempArr = [];
const temp = console.log;
const testLunches = []
try {
  console.log = obj => tempArr.push(obj);
  addLunchToEnd(testLunches, "Tacos");
  assert.strictEqual(tempArr[0], "Tacos added to the end of the lunch menu.");
  addLunchToEnd(testLunches, "Pizza");
  assert.strictEqual(tempArr[1], "Pizza added to the end of the lunch menu.");
} finally {
  console.log = temp;
}
```

`addLunchToEnd(["Pizza", "Tacos"], "Burger")` should return `["Pizza", "Tacos", "Burger"]`.

```js

> assert.deepStrictEqual(addLunchToEnd(["Pizza", "Tacos"], "Burger"), ["Pizza", "Tacos", "Burger"]);
> assert.deepStrictEqual(addLunchToEnd(["Fries"], "Tacos"), ["Fries", "Tacos"]);

Reviewer 2: ```suggestion
const temp = console.log;
console.log = () => {}
try {
  assert.deepStrictEqual(addLunchToEnd(["Pizza", "Tacos"], "Burger"), ["Pizza", "Tacos", "Burger"]);
  assert.deepStrictEqual(addLunchToEnd(["Fries"], "Tacos"), ["Fries", "Tacos"]);
} finally {
  console.log = temp;
}
```
We need to override `console.log` here otherwise campers can see these strings there and it can be confusing.

assert.isFunction(addLunchToStart);
```

The function `addLunchToStart` should have two parameters.

```js
assert.lengthOf(addLunchToStart, 2);
```

`addLunchToStart(lunches, "Sushi")` should log the string `"Sushi added to the start of the lunch menu."` to the console.

```js
const tempArr = [];
const testLunches = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  addLunchToStart(testLunches, "Sushi");
  assert.strictEqual(tempArr[0], "Sushi added to the start of the lunch menu.");  
  addLunchToStart(testLunches, "Burger");
  assert.strictEqual(tempArr[1], "Burger added to the start of the lunch menu.");
} finally {
  console.log = temp;
}
```

`addLunchToStart(["Burger", "Sushi"], "Pizza")` should return `["Pizza", "Burger", "Sushi"]`.

```js

> assert.deepStrictEqual(addLunchToStart(["Burger", "Sushi"], "Pizza"), ["Pizza", "Burger", "Sushi"]);
> assert.deepStrictEqual(addLunchToStart(["Noodles"], "Sushi"), ["Sushi", "Noodles"]);

Reviewer 2: ```suggestion
const temp = console.log;
console.log = () => {}
try {
  assert.deepStrictEqual(addLunchToStart(["Burger", "Sushi"], "Pizza"), ["Pizza", "Burger", "Sushi"]);
  assert.deepStrictEqual(addLunchToStart(["Noodles"], "Sushi"), ["Sushi", "Noodles"]);
} finally {
  console.log = temp;
}
```

assert.isFunction(removeLastLunch)
```

The function `removeLastLunch` should have one parameter.

```js
assert.lengthOf(removeLastLunch, 1);
```

When the input array is empty, the function `removeLastLunch` should log the string `"No lunches to remove."` to the console.

```js
const testLunches = [];
const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  removeLastLunch(testLunches);
  assert.strictEqual(tempArr[0], "No lunches to remove.");
} finally {
  console.log = temp;
}
```

> When `lunches = ["Stew", "Soup", "Toast"]`, `removeLastLunch(lunches)` should log the string `"Toast removed from the end of the lunch menu."` to the console.

Reviewer 2: ```suggestion
`removeLastLunch(["Stew", "Soup", "Toast"])` should log the string `"Toast removed from the end of the lunch menu."` to the console.
```

```js
let testLunches = ["Stew", "Soup", "Toast"];
const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  removeLastLunch(testLunches);
  assert.strictEqual(tempArr[0], "Toast removed from the end of the lunch menu.");
  testLunches = ["Rice", "Pizza"];
  removeLastLunch(testLunches);
  assert.strictEqual(tempArr[1], "Pizza removed from the end of the lunch menu.");
} finally {
  console.log = temp;
}
```

`removeLastLunch(["Sushi", "Pizza", "Noodles"])` should return `["Sushi", "Pizza"]`.

```js

> assert.deepStrictEqual(removeLastLunch(["Sushi", "Pizza", "Noodles"]), ["Sushi", "Pizza"]);
> assert.deepStrictEqual(removeLastLunch(["Pizza", "Burger"]), ["Pizza"]);

Reviewer 2: ```suggestion
const temp = console.log;
console.log = () => {}
try {
  assert.deepStrictEqual(removeLastLunch(["Sushi", "Pizza", "Noodles"]), ["Sushi", "Pizza"]);
  assert.deepStrictEqual(removeLastLunch(["Pizza", "Burger"]), ["Pizza"]);
} finally {
  console.log = temp;
}
```

```

You should define a function `removeFirstLunch`.

```js
assert.isFunction(removeFirstLunch);
```

The function `removeFirstLunch` should have a single parameter.

```js
assert.lengthOf(removeFirstLunch, 1);
```

When the input array is empty, the function `removeFirstLunch` should log the string `"No lunches to remove."` to the console.

```js
const testLunches = [];
const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  removeFirstLunch(testLunches);
  assert.strictEqual(tempArr[0], "No lunches to remove.");
} finally {
  console.log = temp;
}
```

> When `lunches = ["Salad", "Eggs", "Cheese"]`, `removeFirstLunch(lunches)` should log the string `"Salad removed from the start of the lunch menu."` to the console.

Reviewer 2: ```suggestion
`removeFirstLunch(["Salad", "Eggs", "Cheese"])` should log the string `"Salad removed from the start of the lunch menu."` to the console.
```

```js
let testLunches = ["Salad", "Eggs", "Cheese"];
const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  removeFirstLunch(testLunches);
  assert.strictEqual(tempArr[0], "Salad removed from the start of the lunch menu.");
  testLunches = ["Stew", "Soup", "Toast"]
  removeFirstLunch(testLunches);
  assert.strictEqual(tempArr[1], "Stew removed from the start of the lunch menu.");
} finally {
  console.log = temp;
}
```

`removeFirstLunch(["Sushi", "Pizza", "Burger"])` should return `["Pizza", "Burger"]`.

```js

> assert.deepStrictEqual(removeFirstLunch(["Sushi", "Pizza", "Burger"]), ["Pizza", "Burger"]);
> assert.deepStrictEqual(removeFirstLunch(["Pizza", "Burger"]), ["Burger"]);

Reviewer 2: ```suggestion
const temp = console.log;
console.log = () => {}
try {
  assert.deepStrictEqual(removeFirstLunch(["Sushi", "Pizza", "Burger"]), ["Pizza", "Burger"]);
  assert.deepStrictEqual(removeFirstLunch(["Pizza", "Burger"]), ["Burger"]);
} finally {
  console.log = temp;
}
```

```

You should define a function `getRandomLunch`.

```js
assert.isFunction(getRandomLunch);
```

> The function `getRandomLunch` should take a single argument.

Reviewer: ```suggestion
The function `getRandomLunch` should have a single parameter.
```

```js
assert.lengthOf(getRandomLunch, 1);
```

When the input array is empty, the function `getRandomLunch` should log the string `"No lunches available."` to the console.

```js
const testLunches = [];
const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  getRandomLunch(testLunches);
  assert.strictEqual(tempArr[0], "No lunches available.");
} finally {
  console.log = temp;
}
```

> When `lunches = ["Fish", "Fries", "Roast"]`, the function `getRandomLunch` should log a string in the format `"Randomly selected lunch: [Lunch Item]"` to the console.

Reviewer 2: ```suggestion
When the input array is not empty, the function `getRandomLunch` should log a string in the format `"Randomly selected lunch: [Lunch Item]"` to the console.
```

const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  removeFirstLunch(testLunches);
  assert.strictEqual(tempArr[0], "No lunches to remove.");
} finally {
  console.log = temp;
}
```

`removeFirstLunch(["Salad", "Eggs", "Cheese"])` should log the string `"Salad removed from the start of the lunch menu."` to the console.

```js
let testLunches = ["Salad", "Eggs", "Cheese"];
const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  removeFirstLunch(testLunches);
  assert.strictEqual(tempArr[0], "Salad removed from the start of the lunch menu.");
  testLunches = ["Stew", "Soup", "Toast"]
  removeFirstLunch(testLunches);
  assert.strictEqual(tempArr[1], "Stew removed from the start of the lunch menu.");
} finally {
  console.log = temp;
}
```

`removeFirstLunch(["Sushi", "Pizza", "Burger"])` should return `["Pizza", "Burger"]`.

```js
const temp = console.log;
console.log = () => {}
try {
  assert.deepStrictEqual(removeFirstLunch(["Sushi", "Pizza", "Burger"]), ["Pizza", "Burger"]);
  assert.deepStrictEqual(removeFirstLunch(["Pizza", "Burger"]), ["Burger"]);
} finally {
  console.log = temp;
}
```

> 

Reviewer 2: One last thing:
````suggestion

`addLunchToEnd`, `addLunchToStart`, `removeLastLunch`, and `removeFirstLunch` should return the same array passed as an argument after updating it.

```js
const temp = console.log;
console.log = () => {}
const testLunches = ["Spam"];
try {
  assert.equal(addLunchToEnd(testLunches, "Pizza"), testLunches);
  assert.equal(addLunchToStart(testLunches, "Caviar"), testLunches);
  assert.equal(removeLastLunch(testLunches), testLunches);
  assert.equal(removeFirstLunch(testLunches), testLunches);
} finally {
  console.log = temp;
}
```

````
We should add this just in case someone returns a copy of the array instead of the original.

assert.isFunction(showLunchMenu);
```

> The function `showLunchMenu` should take a single argument.

Reviewer: ```suggestion
The function `showLunchMenu` should have a single parameter.
```

```js
assert.lengthOf(showLunchMenu, 1);
```

When the input array is empty, the function `showLunchMenu` should log the string `"The menu is empty."` to the console.

```js
const testLunches = [];
const tempArr = [];
const temp = console.log;
try {
  console.log = obj => tempArr.push(obj);
  showLunchMenu(testLunches);
  assert.strictEqual(tempArr[0], "The menu is empty.");
} finally {
  console.log = temp;
}
```

> When `lunches = ["Greens", "Corns", "Beans"]`, the function `showLunchMenu` should log a string in the format `"Menu items: Greens, Corns, Beans"` to the console.

Reviewer 2: ```suggestion
`showLunchMenu(["Greens", "Corns", "Beans"])` should log `"Menu items: Greens, Corns, Beans"` to the console.
```
